### PR TITLE
Castaway/dismiss notifications in channel

### PR DESCRIPTION
**Pull request by castaway**
Store unread notification count per conversation object, assemble a count of those for the sidebar display. Dismiss / set to zero (for that channel) when a user visits a channel that has notifications.

I'm missing a sane way to test this in __tests__ as it requires both User + Notifications objs, any ideas?

### CODE REVIEW

File: assets/components/ChatSidebar.svelte [hunk 1]
```diff
@@ -11,14 +11,14 @@ import {route} from '../store/Route';
 export let transition;
 
 const user = getContext('user');
-const notifications = $user.notifications;
```
Comment (jhthorsen) [inline assets/components/ChatSidebar.svelte:14]: Why did you change this? I think `$user.notifications` was a typo in the original code though. It should just be `user.notifications`.

File: lib/Convos/Core/Backend.pm [hunk 2]
```diff
@@ -75,7 +75,7 @@ sub _setup {
           my ($connection, $target, $msg) = @_;
 
           if ($msg->{highlight} and $target->id and !$target->is_private) {
-            $connection->user->{unread}++;
+            $target->inc_notifications_p->catch(sub { $self->_debug('inc_notifications %s FAIL %s', $target->id, shift) } );
```
Comment (jhthorsen) [inline lib/Convos/Core/Backend.pm:78]: Since the class doesn't really do anything, you can change this to inc_notifications() instead:

```
sub inc_notifications { $_[0]->{notifications}++; $_[0] }
```

File: lib/Convos/Core/Conversation.pm [hunk 3]
```diff
@@ -9,6 +9,7 @@ has name     => sub { Carp::confess('name required in constructor') };
 has password => '';
 has topic    => '';
 has unread   => 0;
+has notifications => 0;
```
Comment (jhthorsen) [inline lib/Convos/Core/Conversation.pm:12]: Is it easier to read `notifications` or `n_notifications` ?

**Review by jhthorsen — CHANGES_REQUESTED** [review]
I think this is a great start! Thank you 👍 

I'm a bit unsure what `updateNotificationCount()` in Chat.svelte does, but I will look closer when I wake up in the morning.

Can you also get rid of the extra code in Convos::Core::User that is not in use anymore? And I guess the API needs some changes as well to actually clear the notifications.

File: assets/components/ChatSidebar.svelte (hunk 1 above)
Comment (castaway) [reply assets/components/ChatSidebar.svelte:14]: Ah I think that one was me trying to fix a different issue, I'll put it back

File: lib/Convos/Core/Conversation.pm (hunk 3 above)
Comment (castaway) [reply lib/Convos/Core/Conversation.pm:12]: Either works - I otherwise pondered "notifications_count" .. to me its somewhat obvious from the default being 0 .. else we'd need to change all the "unread" attributes too.

Comment (jhthorsen) [reply lib/Convos/Core/Conversation.pm:12]: It's obvious here, but not so obvious other places in the code where it is used. Going to resolve this issue though, since "unread" needs changing as well.

**Comment by castaway** [comment]
Updated with API notifications value resetting to 0 at appropriate points.

Not sure I got Mojo's promises quite right tho as I got this:

```
[2021-07-15 17:47:56.06484] [1821] [debug] [dsWZhdKQfyV9] Routing to controller "Convos::Controller::Notifications" and action "read"
[irc-libera] Save success. (/mnt/allthespace/usrsrc/perl/convos/castaway@desert-island.me.uk/irc-libera/connection.json)
[2021-07-15 17:47:56.06832] [1821] [error] Can't locate object method "clone" via package "Mojo::Collection" at /usr/src/perl/libs/convos/lib/perl5/Mojo/Promise.pm line 120.

[2021-07-15 17:47:56.06956] [1821] [debug] [dsWZhdKQfyV9] 500 Internal Server Error (0.007406s, 135.026/s)
```

File: lib/Convos/Controller/Notifications.pm [hunk 4]
```diff
@@ -16,7 +16,12 @@ sub read {
   my $self = shift->openapi->valid_input or return;
   my $user = $self->backend->user        or return $self->reply->errors([], 401);
 
-  return $user->unread(0)->save_p->then(sub { $self->render(openapi => {}) });
+  Mojo::Promise->all($user->connections->map(sub {
+      my $conn = shift;
+      $conn->conversations->map( sub { shift->notifications(0); });
+      $conn->save_p();
+    })
+  )->then(sub { $self->render(openapi => {}) });
```
Comment (jhthorsen) [inline lib/Convos/Controller/Notifications.pm:24]: I don't think you need this complexity, since the connection is saved on interval. (You should be rather unlucky if it doesn't get saved automatically) I suggest something like this:

```
$user->connections->map(sub { shift->conversations->map(notifications => 0) });
$self->render(openapi => {});
```

But the problem with the original code is that `then()` is called on the collection (return value from map) not the promise method all(). Pretty sure this would work:

```
Mojo::Promise->all(
  $user->connections->map(sub {
    my $conn = shift;
    $conn->conversations->map(sub { shift->notifications(0); });
    return $conn->save_p;
  })->each
)->then(sub { $self->render(openapi => {}) });
```

File: lib/Convos/Controller/Conversation.pm [hunk 5]
```diff
@@ -14,6 +14,7 @@ sub mark_as_read {
   }
 
   $conversation->unread(0);
+  $conversation->notifications(0);
```
Comment (jhthorsen) [inline lib/Convos/Controller/Conversation.pm:17]: This can also be chained:

```
$conversation->notifications(0)->unread(0);
```

File: lib/Convos/Core/Backend.pm [hunk 6]
```diff
@@ -75,7 +75,7 @@ sub _setup {
           my ($connection, $target, $msg) = @_;
 
           if ($msg->{highlight} and $target->id and !$target->is_private) {
-            $connection->user->{unread}++;
+            $target->inc_notifications();
```
Comment (jhthorsen) [inline lib/Convos/Core/Backend.pm:78]: 👍

File: lib/Convos/Core/Conversation.pm [hunk 7]
```diff
@@ -35,7 +38,7 @@ sub _calculate_unread_p {
 
 sub TO_JSON {
   my ($self, $persist) = @_;
-  my %json = map { ($_, $self->$_) } qw(frozen name topic unread);
+  my %json = map { ($_, $self->$_) } qw(frozen name topic unread notifications);
```
Comment (jhthorsen) [inline lib/Convos/Core/Conversation.pm:41]: I forgot to mention something else: I would appreciate if "things" where sorted. This goes for methods and attributes as well.

Also, if you could run the code through `perltidy`, then it would be less work afterwards, and you would get mentioned with "git blame".

I'm going to take this change either way, but I would greatly appreciate it.

**Review by jhthorsen — CHANGES_REQUESTED** [review]

File: lib/Convos/Controller/Notifications.pm (hunk 4 above)
Comment (castaway) [reply lib/Convos/Controller/Notifications.pm]: Thanks! Revised as suggested

File: lib/Convos/Core/Conversation.pm (hunk 7 above)
Comment (castaway) [reply lib/Convos/Core/Conversation.pm]: Righto, I shall try and remember that (updated)

File: lib/Convos/Controller/Conversation.pm (hunk 5 above)
Comment (castaway) [reply lib/Convos/Controller/Conversation.pm]: done

**Review by jhthorsen — CHANGES_REQUESTED** [review]
There’s many attributes/methods/keywords that isn’t sorted, and the code has not been run through perltidy. (I use https://metacpan.org/dist/App-githook-perltidy/view/bin/githook-perltidy)

File: assets/page/Chat.svelte [hunk 8]
```diff
@@ -39,6 +39,7 @@ let focusChatInput, fillIn, uploader;
 
 $: setConversationFromRoute(connection_id, conversation_id);
 $: setConversationFromUser($user);
+$: updateNotificationCount($user);
```
[the remark this answers is not in the document]
Comment (castaway) [reply assets/page/Chat.svelte:42]: This is the update that sets notifications to 0 for this channel, when the user visits it (so should be on load of this channel, not on every message)

[the remark this answers is not in the document]
Comment (castaway) [reply assets/page/Chat.svelte:42]: I put it here as that's where setActiveConversation is, which I assumed happened once per channel / page visit.

**Comment by jhthorsen** [comment]
I've rebased your commits and fixed a couple of things. The changes are in 48e904850c5abfb1093ef7098a05c731bfe83a5c and 96ef5a269c2a2c9aaeb92a72af7afcab2045343c
